src/spl/spline.h

```diff
@@ -236,8 +257,113 @@ class Spline {
+        --position;
+        --diff[i - 1];
+      }
+      InsertKnot(GetKnotVector(dimension)->GetKnot(position), dimension, static_cast<size_t>(diff[i - 1]));
+    }
+    return diff;
+  }
+
+  void RemoveBezierKnots(std::vector<int> diff, int dimension) {
+    size_t position = GetKnotVector(dimension)->GetNumberOfKnots() - GetDegree(dimension).get() - 2;
+    for (size_t i = GetKnotVector(dimension)->GetNumberOfDifferentKnots() - 2; i > 0; --i, --position) {
+      while (GetKnotVector(dimension)->GetKnot(position) == GetKnotVector(dimension)->GetKnot(position - 1)) {
+        --position;
+      }
+      RemoveKnot(GetKnotVector(dimension)->GetKnot(position), dimension,
+                 util::NumericSettings<double>::kEpsilon(), static_cast<size_t>(diff[i - 1] - 1));
+    }
+  }
+
+  virtual int GetBezierPointLength() const = 0;
+
+  virtual std::vector<double> GetBezierSegment(int dimension, int segment) const {
+    util::MultiIndexHandler<DIM> point_handler(GetPointsPerDirection());
+    int width = GetDegree(dimension).get() + 1;
+    int segment_length = GetNumberOfControlPoints() / GetPointsPerDirection()[dimension];
+    int point_length = this->GetBezierPointLength();
+    std::vector<double> bezier_cps(static_cast<size_t>(point_length * width * segment_length));
+    for (int i = 0; i < point_handler.Get1DLength(); ++i, ++point_handler) {
+      auto index_in_dir = point_handler[dimension];
+      if (index_in_dir >= segment * width && index_in_dir < (segment + 1) * width) {
+        double weight = this->GetWeight(point_handler.GetIndices());
+        if (point_length == GetPointDim() + 1) {
+          auto weight_index = (index_in_dir % width + point_handler.ExtractDimension(dimension) * width) * point_length
+              + this->GetPointDim();
+          bezier_cps[weight_index] = weight;
+        }
+        for (int j = 0; j < GetPointDim(); ++j) {
+          auto index = (index_in_dir % width + point_handler.ExtractDimension(dimension) * width) * point_length + j;
+          bezier_cps[index] = GetControlPoint(point_handler.GetIndices(), j) * weight;
+        }
+      }
+    }
+    return bezier_cps;
+  }
+
+  std::vector<std::pair<baf::ControlPoint, double>>
+  DegreeElevateBezierSegment(std::vector<double> bezier_cps, std::vector<double> alpha, int dimension) const {
+    int width = GetDegree(dimension).get() + 1;
+    int segment_length = GetNumberOfControlPoints() / GetPointsPerDirection()[dimension];
+    util::MultiIndexHandler<2> point_handler({width, segment_length});
+    int point_length = this->GetBezierPointLength();
+    std::vector<std::pair<baf::ControlPoint, double>> weighted_cps;
+    std::vector<double> coord(static_cast<size_t>(GetPointDim()));
+    for (int k = 0; k < point_handler.Get1DLength(); ++k, ++point_handler) {
+      auto index = point_handler[0];
+      auto pos = point_handler.Get1DIndex();
+      double new_weight = point_length == GetPointDim() ? 1.0 :
```

Comment:
Replace ternary operator by if-then-else.

Reply:
Then shouldn't there be different implementations for non-rational & rational splines of the whole method instead?

Reply:
At a first glance it seems to me as if there are common operations for B-splines & NURBS during degree elevation. Thus, this functionality should be gathered in the base class `spline`. However, calls to functions with different implementations for different spline types are always possible. In this case, a virtual method like `ComputeWeightForDegreeElevation` might be appropriate?